app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/app/tracking/AppTPVpnConnectivityLossListener.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.mobile.android.app.tracking
+
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.di.scopes.VpnScope
+import com.duckduckgo.mobile.android.vpn.feature.AppTpFeatureConfig
+import com.duckduckgo.mobile.android.vpn.feature.AppTpSetting
+import com.duckduckgo.mobile.android.vpn.service.VpnServiceCallbacks
+import com.duckduckgo.mobile.android.vpn.service.connectivity.VpnConnectivityLossListenerPlugin
+import com.duckduckgo.mobile.android.vpn.state.VpnStateMonitor
+import com.duckduckgo.networkprotection.api.NetworkProtectionState
+import com.squareup.anvil.annotations.ContributesMultibinding
+import dagger.SingleInstanceIn
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import logcat.LogPriority
+import logcat.logcat
+
+@ContributesMultibinding(
+    scope = VpnScope::class,
+    boundType = VpnConnectivityLossListenerPlugin::class,
+)
+@ContributesMultibinding(
+    scope = VpnScope::class,
+    boundType = VpnServiceCallbacks::class,
+)
+@SingleInstanceIn(VpnScope::class)
+class AppTPVpnConnectivityLossListener @Inject constructor(
+    private val networkProtectionState: NetworkProtectionState,
+    private val appTrackingProtection: AppTrackingProtection,
+    private val appTpFeatureConfig: AppTpFeatureConfig,
+    private val dispatcherProvider: DispatcherProvider,
+) : VpnConnectivityLossListenerPlugin, VpnServiceCallbacks {
+
+    private var connectivityLostEvents = 0
+
+    override fun onVpnConnected(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
+            if (isActive()) {
+                connectivityLostEvents = 0
+            }
+        }
+    }
+
+    override fun onVpnConnectivityLoss(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
+            if (isActive() && connectivityLostEvents++ > 1) {
+                logcat(LogPriority.WARN) { "AppTP detected connectivity loss, re-configuring..." }
+                connectivityLostEvents = 0
+                appTrackingProtection.restart()
```

Comment:
Should I do `Runtime.getRuntime().exit(0)` instead?

Reply:
What advantage to we get from killing the process? Wouldn't this potentially cause last state to be DISABLED with unknown reason which will cause always on to show more?

Reply:
Kill process would ensure all state is freed, for instance memory. We don't have evidence of eg. mem leaks, but if we did process kill would get rid of them.

Another Q, do you think we should put a limit to re-tries? ie. if we try to reconnect 3-4 times without success we basically give up and let user re-enable? If we can recover, it may be indicative of some issue we can't really solve without stopping the VPN service (?)